### PR TITLE
Allow custom range sequence in StronglyEntanglingLayers

### PR DESCRIPTION
**Pull request by wongwsvincent**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -242,6 +242,9 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 * Prevent Hamiltonians that share wires from being multiplied together.
   [(#1273)](https://github.com/PennyLaneAI/pennylane/pull/1273)
 
+* Fixed the custom range sequence feature and added test for it.
+  [(#1332)](https://github.com/PennyLaneAI/pennylane/pull/1332)
```
Comment (josh146) [inline .github/CHANGELOG.md:246]: It would be good to provide a few more details here, for example
```suggestion
* Fixed a bug where the custom range sequences could not be passed
  to the `StronglyEntanglingLayers` template.
  [(#1332)](https://github.com/PennyLaneAI/pennylane/pull/1332)
```

**Review by josh146 — COMMENTED** [review]
Thanks for this nice contribution @wongwsvincent! Is this PR now ready for review?

**Comment by wongwsvincent** [comment]
> Thanks for this nice contribution @wongwsvincent! Is this PR now ready for review?

Hi @josh146 , I am having trouble not decreasing the coverage of the tests. I included a ```raise ValueError``` function to prevent the user from using range values equal to zero or divisible by the number of wires, but that line of code would never be reached by the tests. I tried including a bad parameter test case that would cause the error, but that would fail the core-tests as expected. I have just reverted back to the previous version by removing that bad test case. Let me know if there's something I can try to avoid the decrease in coverage; otherwise, this PR is ready for review.

File: pennylane/templates/layers/strongly_entangling.py [hunk 2]
```diff
@@ -92,6 +92,17 @@ def __init__(self, weights, wires, ranges=None, imprimitive=None, do_queue=True)
                 self.ranges = [(l % (len(wires) - 1)) + 1 for l in range(self.n_layers)]
             else:
                 self.ranges = [0] * self.n_layers
+        else:
+            for r in ranges:
+                if r % len(wires) == 0:
+                    raise ValueError(
+                        f"Rangers must not be zero nor divisible by the number of wires; got {r}"
```
Comment (mariaschuld) [inline pennylane/templates/layers/strongly_entangling.py:99]:
```suggestion
                        f"Ranges must not be zero nor divisible by the number of wires; got {r}"
```

File: pennylane/templates/layers/strongly_entangling.py [hunk 3]
```diff
@@ -92,6 +92,17 @@ def __init__(self, weights, wires, ranges=None, imprimitive=None, do_queue=True)
...
+                    )
+            self.ranges = ranges
+            # repeat range sequence if the length is shorter than the number of layers
+            if len(ranges) < self.n_layers:
+                for i in range(self.n_layers - len(ranges)):
+                    self.ranges.append(self.ranges[i])
```
Comment (mariaschuld) [inline pennylane/templates/layers/strongly_entangling.py:105]:
```suggestion
```
I would rather not fill the `ranges` up; if a user decides to fine-tune this hyperparameter the onus is on her to make sure it has the correct size...which also gives better control about what is happening in the template. What do you think?

We could add a check that `len(ranges) == len(self.n_layers)`?

File: pennylane/templates/layers/strongly_entangling.py [hunk 4]
```diff
@@ -92,6 +92,17 @@ def __init__(self, weights, wires, ranges=None, imprimitive=None, do_queue=True)
                 self.ranges = [(l % (len(wires) - 1)) + 1 for l in range(self.n_layers)]
             else:
                 self.ranges = [0] * self.n_layers
+        else:
+            for r in ranges:
```
Comment (mariaschuld) [inline pennylane/templates/layers/strongly_entangling.py:96]: I am on the fence about this check...it is always tricky to get the right balance of slowdown due to checks vs. help for the user to avoid wrong inputs. Happy to leave it in though. 

If you want, one could add a sentence to the docstrings where `ranges` is described, saying that it cannot be zero or M, since that would result in one wire for the 2-qubit gate.

File: tests/templates/test_layers/test_strongly_entangling.py [hunk 5]
```diff
@@ -89,6 +89,30 @@ def circuit2():
...
+        op = qml.templates.StronglyEntanglingLayers(
+            weights=weights, wires=range(n_wires), ranges=ranges
+        )
+        ops = op.expand().operations
+
+        gate_wires = [gate.wires for gate in ops]
+        range_idx = 0
+        for idx, i in enumerate(gate_wires):
+            if idx % (n_wires * 2) // n_wires == 1:
+                test_wire = qml.wires.Wires(
```
Comment (mariaschuld) [inline tests/templates/test_layers/test_strongly_entangling.py:109]: Nitpicking, but one could also call this `expected` or `expected_wire`...:)

File: tests/templates/test_layers/test_strongly_entangling.py [hunk 6]
```diff
@@ -89,6 +89,30 @@ def circuit2():
...
+        """Test that correct sequence of custom ranges are used in the circuit."""
+
+        weights = np.random.randn(n_layers, n_wires, 3)
+
+        op = qml.templates.StronglyEntanglingLayers(
+            weights=weights, wires=range(n_wires), ranges=ranges
+        )
+        ops = op.expand().operations
+
+        gate_wires = [gate.wires for gate in ops]
```
Comment (mariaschuld) [inline tests/templates/test_layers/test_strongly_entangling.py:105]: You could extract `gate.wires.labels` here, and then use a tuple for the `test_wire`, so you don't need the `Wires` class below...

File: tests/templates/test_layers/test_strongly_entangling.py [hunk 7]
```diff
@@ -89,6 +89,30 @@ def circuit2():
...
+        range_idx = 0
+        for idx, i in enumerate(gate_wires):
+            if idx % (n_wires * 2) // n_wires == 1:
+                test_wire = qml.wires.Wires(
+                    [0 + idx % n_wires, (ranges[range_idx % len(ranges)] + idx % n_wires) % n_wires]
+                )
+                assert i == test_wire
+                if idx % n_wires == n_wires - 1:
+                    range_idx += 1
```
Comment (mariaschuld) [inline tests/templates/test_layers/test_strongly_entangling.py:114]: Wow, that quite some involved logic here to pick out the right wires :)

**Review by mariaschuld — CHANGES_REQUESTED** [review]
Thanks so much for this contribution @wongwsvincent, this is great!

I left a few comments, but most are just nitpicking, and I am happy to be convinced otherwise...

About the coverage: you can (and should) test all errors with the context  `with pytest.raises(ValueError, match="<part_of_the_error_message>"):`, in which you can create the op. If the error is not raised, the test will fail. Would that help, or did I misunderstand your problem?

File: pennylane/templates/layers/strongly_entangling.py (hunk 3 above)
Comment (wongwsvincent) [reply pennylane/templates/layers/strongly_entangling.py]: Okay, you have me convinced. The user should be responsible for it. The function does look cleaner that way by being simpler and straightforward. I will add a check as you suggested.

File: pennylane/templates/layers/strongly_entangling.py (hunk 4 above)
Comment (wongwsvincent) [reply pennylane/templates/layers/strongly_entangling.py:96]: I tend to keep this check in. Otherwise, the runtime error message would not be very informative about what went wrong (or sometimes just stuck on a loop without error message if one is only trying to draw the circuit). Also, this check only happens on initialization and costs computation time of O(number of layers), I believe the cost would not be noticeable.

File: tests/templates/test_layers/test_strongly_entangling.py (hunk 5 above)
Comment (wongwsvincent) [reply tests/templates/test_layers/test_strongly_entangling.py:109]: Better clarity is good. I will change it to ```expected_wire```

**Comment by wongwsvincent** [comment]
Thanks so much for reviewing the code, @mariaschuld .
And thanks for the tips of using ```pytest.raises```. It sounds exactly what I am looking for! I will add the error check in to maintain good coverage.

File: pennylane/templates/layers/strongly_entangling.py (hunk 4 above)
Comment (josh146) [reply pennylane/templates/layers/strongly_entangling.py:100]: @wongwsvincent something to keep in mind is that because of PennyLane's functional user-interface, initialization will occur every time that the QNode is executed 🙂 So something to keep in mind --- overheads can add up quickly.

(Long term, a strategy where we 'just in time compile' the function would help eliminate this overhead)

File: tests/templates/test_layers/test_strongly_entangling.py [hunk 8]
```diff
@@ -116,6 +141,14 @@ def circuit(weights):
...
             circuit(weights)
 
+        with pytest.raises(ValueError, match="Range sequence must be of length"):
+            weights = np.random.randn(2, 2, 3)
+            circuit(weights, ranges=[1])
+
+        with pytest.raises(ValueError, match="Ranges must not be zero nor"):
+            weights = np.random.randn(1, 2, 3)
+            circuit(weights, ranges=[0])
+
```
Comment (mariaschuld) [inline tests/templates/test_layers/test_strongly_entangling.py:151]: Sorry to be a pain - would you mind moving this to a separate test function? Now it is under "test that number of dimensions are correct" which is confusing. You could just add a function `test_exception_wrong_ranges` or so...

Comment (mariaschuld) [reply tests/templates/test_layers/test_strongly_entangling.py:151]: Otherwise, this is great!

File: pennylane/templates/layers/strongly_entangling.py (hunk 4 above)
Comment (wongwsvincent) [reply pennylane/templates/layers/strongly_entangling.py:100]: @josh146 Ah ha, thanks a lot for the explanation! I feel like we should keep it in to return a more helpful error message. Should we leave it in for this round, and try to eliminate the overhead in the future?

File: tests/templates/test_layers/test_strongly_entangling.py (hunk 8 above)
Comment (wongwsvincent) [reply tests/templates/test_layers/test_strongly_entangling.py:151]: Totally! Sorry for overlooking it. I can do so quickly.

File: pennylane/templates/layers/strongly_entangling.py (hunk 4 above)
Comment (mariaschuld) [reply pennylane/templates/layers/strongly_entangling.py:100]: I benchmarked with and without checks and the difference is negligible (on average about 21.4s vs 21.3s, but fluctuations of repeating the benchmark on my machine vary that first decimal digit quite a bit). And @wongwsvincent is right, the error message without checks is horrendous.

Since not many users will use custom ranges, I propose we leave both checks in!

**Review by mariaschuld — CHANGES_REQUESTED** [review]
Hey @wongwsvincent .

Thanks for updating the PR - you notice that we are quite strict with code review :) 

The only last change would be to move the error tests into a separate test function, then it's ready to be merged from my side! Just ask for a new review when this is done.

Oh, and don't forget to add yourself to the list of contributors in the CHANGELOG (at the bottom of the new version's release notes) if you're not already there!

**Comment by wongwsvincent** [comment]
I actually appreciate the strict code review a lot. It's my first time contributing and I definitely learned a lot =D 

I have moved the check of error raising regarding the range values to a separate test function. And gratefully added my name in the CHANGELOG

Code should be ready for re-review @mariaschuld

File: pennylane/templates/layers/strongly_entangling.py (hunk 4 above)
Comment (wongwsvincent) [reply pennylane/templates/layers/strongly_entangling.py:100]: Thanks so much for checking the computation time quickly @mariaschuld !
